File: shared/logging/go/logger.go
// Package logging provides structured JSON logging for AIAUDITOR services.
// Log format matches tech spec §10.1:
//
//	{timestamp, level, service, trace_id, span_id, org_id, engagement_id, event, metadata}
//
// Usage:
//
//	logger := logging.New("my-service")
//	logger.Info(ctx, "user.login", map[string]any{"user_id": id})
package logging

import (
	"context"
	"encoding/json"
	"io"
	"os"
	"time"
)

// Level is a log severity level.
type Level string

const (
	LevelInfo     Level = "INFO"
	LevelWarn     Level = "WARN"
	LevelError    Level = "ERROR"
	LevelCritical Level = "CRITICAL"
)

// contextKey is an unexported type for context keys owned by this package.
type contextKey string

const (
	keyTraceID      contextKey = "trace_id"
	keySpanID       contextKey = "span_id"
	keyOrgID        contextKey = "org_id"
	keyEngagementID contextKey = "engagement_id"
	keyRequestID    contextKey = "request_id"
)

// Entry is a single structured log entry matching tech spec §10.1.
type Entry struct {
	Timestamp    string         `json:"timestamp"`
	Level        Level          `json:"level"`
	Service      string         `json:"service"`
	TraceID      string         `json:"trace_id,omitempty"`
	SpanID       string         `json:"span_id,omitempty"`
	OrgID        string         `json:"org_id,omitempty"`
	EngagementID string         `json:"engagement_id,omitempty"`
	Event        string         `json:"event"`
	Metadata     map[string]any `json:"metadata,omitempty"`
}

// Logger emits structured JSON log entries to an io.Writer (default: os.Stdout).
type Logger struct {
	service string
	out     io.Writer
}

// New creates a Logger for the named service.
func New(service string) *Logger {
	return &Logger{service: service, out: os.Stdout}
}

// WithWriter returns a new Logger that writes to w. Useful for testing.
func (l *Logger) WithWriter(w io.Writer) *Logger {
	return &Logger{service: l.service, out: w}
}

// emit serialises and writes one log entry.
func (l *Logger) emit(ctx context.Context, level Level, event string, metadata map[string]any) {
	entry := Entry{
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
		Level:     level,
		Service:   l.service,
		Event:     event,
		Metadata:  metadata,
	}
	if ctx != nil {
		if v, ok := ctx.Value(keyTraceID).(string); ok {
			entry.TraceID = v
		}
		if v, ok := ctx.Value(keySpanID).(string); ok {
			entry.SpanID = v
		}
		if v, ok := ctx.Value(keyOrgID).(string); ok {
			entry.OrgID = v
		}
		if v, ok := ctx.Value(keyEngagementID).(string); ok {
			entry.EngagementID = v
		}
	}
	b, err := json.Marshal(entry)
	if err != nil {
		return
	}
	b = append(b, '\n')
	_, _ = l.out.Write(b)
}

// Info logs at INFO level.
func (l *Logger) Info(ctx context.Context, event string, metadata map[string]any) {
	l.emit(ctx, LevelInfo, event, metadata)
}

// Warn logs at WARN level.
func (l *Logger) Warn(ctx context.Context, event string, metadata map[string]any) {
	l.emit(ctx, LevelWarn, event, metadata)
}

// Error logs at ERROR level.
func (l *Logger) Error(ctx context.Context, event string, metadata map[string]any) {
	l.emit(ctx, LevelError, event, metadata)
}

// Critical logs at CRITICAL level.
func (l *Logger) Critical(ctx context.Context, event string, metadata map[string]any) {
	l.emit(ctx, LevelCritical, event, metadata)
}

// --- Context helpers --------------------------------------------------------

// WithTraceID stores a trace ID in the context.
func WithTraceID(ctx context.Context, traceID string) context.Context {
	return context.WithValue(ctx, keyTraceID, traceID)
}

// WithSpanID stores a span ID in the context.
func WithSpanID(ctx context.Context, spanID string) context.Context {
	return context.WithValue(ctx, keySpanID, spanID)
}

// WithOrgID stores an org ID in the context.
func WithOrgID(ctx context.Context, orgID string) context.Context {
	return context.WithValue(ctx, keyOrgID, orgID)
}

// WithEngagementID stores an engagement ID in the context.
func WithEngagementID(ctx context.Context, engagementID string) context.Context {
	return context.WithValue(ctx, keyEngagementID, engagementID)
}

// TraceIDFromContext extracts the trace ID from ctx, returning "" if absent.
func TraceIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keyTraceID).(string)
	return v
}

// SpanIDFromContext extracts the span ID from ctx, returning "" if absent.
func SpanIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keySpanID).(string)
	return v
}

// OrgIDFromContext extracts the org ID from ctx.
func OrgIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keyOrgID).(string)
	return v
}

// EngagementIDFromContext extracts the engagement ID from ctx.
func EngagementIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keyEngagementID).(string)
	return v
}

// RequestIDFromContext extracts the request ID from ctx.
func RequestIDFromContext(ctx context.Context) string {
	v, _ := ctx.Value(keyRequestID).(string)
	return v
}

File: frontend/src/lib/logger.ts
/**
 * Structured JSON logger for AIAUDITOR frontend API routes.
 *
 * Emits log entries matching tech spec §10.1:
 *   {timestamp, level, service, trace_id, span_id, org_id, engagement_id, event, metadata}
 *
 * Usage:
 *   import { createLogger } from '@/lib/logger'
 *   const log = createLogger('frontend')
 *   log.info(ctx, 'user.login', { user_id: id })
 */

import type { NextRequest } from 'next/server'

/** W3C Trace Context header names. */
export const HEADER_TRACEPARENT = 'traceparent'
export const HEADER_TRACESTATE = 'tracestate'
export const HEADER_REQUEST_ID = 'x-request-id'

/** Log severity levels matching tech spec §10.1. */
export type LogLevel = 'INFO' | 'WARN' | 'ERROR' | 'CRITICAL'

/** A single structured log entry matching tech spec §10.1. */
export interface LogEntry {
  timestamp: string
  level: LogLevel
  service: string
  trace_id?: string
  span_id?: string
  org_id?: string
  engagement_id?: string
  event: string
  metadata?: Record<string, unknown>
}

/** LogContext carries trace and tenant IDs extracted from a request. */
export interface LogContext {
  traceId?: string
  spanId?: string
  orgId?: string
  engagementId?: string
  requestId?: string
}

/** Parse a W3C traceparent header into {traceId, spanId}. */
export function parseTraceparent(
  header: string | null | undefined
): { traceId: string; spanId: string } | null {
  if (!header) return null
  const parts = header.split('-')
  if (parts.length !== 4) return null
  const [version, traceId, spanId, flags] = parts
  if (version !== '00' || traceId.length !== 32 || spanId.length !== 16 || flags.length !== 2) {
    return null
  }
  return { traceId, spanId }
}

/** Format a W3C traceparent header value. */
export function formatTraceparent(traceId: string, spanId: string): string {
  return `00-${traceId}-${spanId}-01`
}

/** Extract LogContext from a Next.js API route request. */
export function contextFromRequest(req: NextRequest): LogContext {
  const traceparent = req.headers.get(HEADER_TRACEPARENT)
  const parsed = parseTraceparent(traceparent)
  return {
    traceId: parsed?.traceId,
    spanId: parsed?.spanId,
    orgId: req.headers.get('x-org-id') ?? undefined,
    requestId: req.headers.get(HEADER_REQUEST_ID) ?? undefined,
  }
}

/** A typed structured logger bound to a service name. */
export class Logger {
  constructor(private readonly service: string) {}

  private emit(
    level: LogLevel,
    ctx: LogContext,
    event: string,
    metadata?: Record<string, unknown>
  ): void {
    const entry: LogEntry = {
      timestamp: new Date().toISOString(),
      level,
      service: this.service,
      event,
    }
    if (ctx.traceId) entry.trace_id = ctx.traceId
    if (ctx.spanId) entry.span_id = ctx.spanId
    if (ctx.orgId) entry.org_id = ctx.orgId
    if (ctx.engagementId) entry.engagement_id = ctx.engagementId
    if (metadata && Object.keys(metadata).length > 0) entry.metadata = metadata

    // In production Next.js, console.log goes to stdout as JSON.
    // eslint-disable-next-line no-console
    console.log(JSON.stringify(entry))
  }

  info(ctx: LogContext, event: string, metadata?: Record<string, unknown>): void {
    this.emit('INFO', ctx, event, metadata)
  }

  warn(ctx: LogContext, event: string, metadata?: Record<string, unknown>): void {
    this.emit('WARN', ctx, event, metadata)
  }

  error(ctx: LogContext, event: string, metadata?: Record<string, unknown>): void {
    this.emit('ERROR', ctx, event, metadata)
  }

  critical(ctx: LogContext, event: string, metadata?: Record<string, unknown>): void {
    this.emit('CRITICAL', ctx, event, metadata)
  }
}

/** Create a structured logger bound to the given service name. */
export function createLogger(service: string): Logger {
  return new Logger(service)
}

/** Default frontend logger instance. */
export const logger = createLogger('frontend')

